File: app/src/main/java/edu/gatech/spellastory/data/Phonemes.java
package edu.gatech.spellastory.data;

import com.opencsv.CSVReader;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.Reader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import edu.gatech.spellastory.domain.Phoneme;

public class Phonemes {

    // Maps phoneme number to list of letter combos
    // Ex. 1 -> [a], 84 -> [s,ss]
    private Map<String, List<String>> phonemes;

    public Phonemes(Reader csvReader) {
        phonemes = readPhonemes(csvReader);
    }

    private Map<String, List<String>> readPhonemes(Reader csvReader) {
        try {
            Map<String, List<String>> phonemes = new HashMap<>();
            CSVReader reader = new CSVReader(csvReader);
            String[] nextLine;
            while ((nextLine = reader.readNext()) != null) {
                String code = nextLine[0];
                String[] letters = Arrays.copyOfRange(nextLine, 1, nextLine.length);
                phonemes.put(code, Arrays.asList(letters));
            }
            return phonemes;
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    List<String> getPhonemeSpellings(String codedPhoneme) {
        if (codedPhoneme.equals("0")) {
            return new ArrayList<>(); // 0 is a silent phoneme
        }
        return phonemes.get(codedPhoneme);
    }

    public List<Phoneme> getAllPhonemes() {
        List<Phoneme> allSpellings = new ArrayList<>();
        for (Map.Entry<String, List<String>> entry : phonemes.entrySet()) {
            for (String spelling : entry.getValue()) {
                allSpellings.add(new Phoneme(entry.getKey(), spelling));
            }
        }
        return allSpellings;
    }
}

File: app/src/main/java/edu/gatech/spellastory/data/Phonemes.kt
package edu.gatech.spellastory.data

import android.app.Activity
import android.content.Context
import com.opencsv.CSVIterator
import com.opencsv.CSVReader
import edu.gatech.spellastory.domain.Phoneme
import edu.gatech.spellastory.util.SingletonHolder
import java.io.InputStreamReader
import java.util.*

val Activity.PhonemesDb
    get() = Phonemes.getInstance(this)

class Phonemes private constructor(context: Context) {

    private val phonemes: Map<String, List<String>> by lazy {
        val reader = CSVReader(InputStreamReader(context.assets.open("coded_phonemes.csv")))
        val phonemes = read(reader)
        reader.close()
        phonemes
    }

    private fun read(reader: CSVReader) = CSVIterator(reader).asSequence().map { line ->
        val code = line[0]
        val letters = Arrays.copyOfRange<String>(line, 1, line.size)
        code to letters.toList()
    }.toMap()

    fun getPhonemeSpellings(code: String): List<String> {
        if (code == "0") return listOf("")
        return phonemes.getValue(code)
    }

    fun getPhonemes(codes: Array<out String>): List<Phoneme> {
        val phonemes = mutableListOf<Phoneme>()
        codes.forEach { code ->
            getPhonemeSpellings(code).sortedByDescending { it.length }.forEach { spelling ->
                phonemes.add(Phoneme(code, spelling))
            }
        }
        return phonemes
    }

    // fun getAllPhonemes(): List<Phoneme> {
    //     val phonemes = mutableListOf<Phoneme>()
    //     this.phonemes.entries.forEach {
    //         val code = it.key
    //         val spellings = it.value
    //         spellings.forEach { spelling ->
    //             phonemes.add(Phoneme(code, spelling))
    //         }
    //     }
    //     return phonemes
    // }

    val randomPhoneme: Phoneme
        get() {
            val code = phonemes.keys.random()
            val spelling = phonemes.getValue(code).random()
            return Phoneme(code, spelling)
        }

    companion object : SingletonHolder<Phonemes, Context>(::Phonemes)
}